File: android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/ApplicationUtils.java
package com.appcoins.sdk.billing;

import static com.appcoins.sdk.core.logger.Logger.logDebug;
import static com.appcoins.sdk.billing.utils.AppcoinsBillingConstants.RESPONSE_CODE;
import static com.appcoins.sdk.core.logger.Logger.logError;

import android.app.Activity;
import android.content.Intent;
import android.util.Base64;
import com.appcoins.sdk.billing.analytics.SdkAnalytics;
import com.appcoins.sdk.billing.helpers.WalletUtils;
import com.appcoins.sdk.billing.listeners.SDKWebResponse;
import com.appcoins.sdk.billing.usecases.mmp.SendSuccessfulPurchaseResponseEvent;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import org.json.JSONObject;

class ApplicationUtils {

  private static final String RESPONSE_INAPP_PURCHASE_DATA = "INAPP_PURCHASE_DATA";
  private static final String RESPONSE_INAPP_SIGNATURE = "INAPP_DATA_SIGNATURE";
  private static final String RESPONSE_INAPP_PURCHASE_ID = "INAPP_PURCHASE_ID";

  static void handleActivityResult(Billing billing, int resultCode, Intent data,
                                   PurchasesUpdatedListener purchaseFinishedListener) {

    SdkAnalytics sdkAnalytics = WalletUtils.getSdkAnalytics();

    if (data == null) {
      logError("Null data in IAB activity result.");
      purchaseFinishedListener.onPurchasesUpdated(ResponseCode.ERROR.getValue(),
          Collections.emptyList());
      return;
    }

    int responseCode = getResponseCodeFromIntent(data);
    String purchaseData = data.getStringExtra(RESPONSE_INAPP_PURCHASE_DATA);
    String dataSignature = data.getStringExtra(RESPONSE_INAPP_SIGNATURE);

    if (resultCode == Activity.RESULT_OK && responseCode == ResponseCode.OK.getValue()) {
      sdkAnalytics.sendPurchaseStatusEvent("success", getResponseDesc(responseCode));
      logDebug("Successful resultcode from purchase activity.");
      logDebug("Purchase data: " + purchaseData);
      logDebug("Data signature: " + dataSignature);
      logDebug("Extras: " + data.getExtras());

      if (purchaseData == null || dataSignature == null) {
        logError("BUG: either purchaseData or dataSignature is null.");
        logDebug("Extras: " + data.getExtras());
        purchaseFinishedListener.onPurchasesUpdated(ResponseCode.ERROR.getValue(),
            Collections.emptyList());

        return;
      }

      if (billing.verifyPurchase(purchaseData, Base64.decode(dataSignature, Base64.DEFAULT))) {
        JSONObject purchaseDataJSON;
        try {
          purchaseDataJSON = new JSONObject(purchaseData);
          Purchase purchase =
              new Purchase(getObjectFromJson(purchaseDataJSON, "orderId"), "inapp", purchaseData,
                  Base64.decode(dataSignature, Base64.DEFAULT),
                  Long.parseLong(getObjectFromJson(purchaseDataJSON, "purchaseTime")),
                  Integer.decode(getObjectFromJson(purchaseDataJSON, "purchaseState")),
                  getObjectFromJson(purchaseDataJSON, "developerPayload"),
                  getObjectFromJson(purchaseDataJSON, "purchaseToken"),
                  getObjectFromJson(purchaseDataJSON, "packageName"),
                  getObjectFromJson(purchaseDataJSON, "productId"),
                  Boolean.parseBoolean(getObjectFromJson(purchaseDataJSON, "isAutoRenewing")));
          List<Purchase> purchases = new ArrayList<>();
          purchases.add(purchase);
          SendSuccessfulPurchaseResponseEvent.Companion.invoke(purchase);
          purchaseFinishedListener.onPurchasesUpdated(responseCode, purchases);
        } catch (Exception e) {
          e.printStackTrace();
          purchaseFinishedListener.onPurchasesUpdated(ResponseCode.ERROR.getValue(),
              Collections.emptyList());
          logError("Failed to parse purchase data.");
        }
      } else {
        logError("Signature verification failed for sku:");
        purchaseFinishedListener.onPurchasesUpdated(ResponseCode.ERROR.getValue(),
            Collections.emptyList());
      }
    } else if (resultCode == Activity.RESULT_OK) {
      // result code was OK, but in-app billing response was not OK.
      logDebug("Result code was OK but in-app billing response was not OK: " + getResponseDesc(
          responseCode));
      sdkAnalytics.sendPurchaseStatusEvent("error", getResponseDesc(responseCode));
      purchaseFinishedListener.onPurchasesUpdated(responseCode, Collections.emptyList());
    } else if (resultCode == Activity.RESULT_CANCELED) {
      logDebug("Purchase canceled - Response: " + getResponseDesc(responseCode));
      sdkAnalytics.sendPurchaseStatusEvent("user_canceled", getResponseDesc(responseCode));
      purchaseFinishedListener.onPurchasesUpdated(ResponseCode.USER_CANCELED.getValue(),
          Collections.emptyList());
    } else {
      logError("Purchase failed. Result code: " + resultCode + ". Response: " + getResponseDesc(
          responseCode));
      sdkAnalytics.sendPurchaseStatusEvent("error", getResponseDesc(responseCode));
      purchaseFinishedListener.onPurchasesUpdated(ResponseCode.ERROR.getValue(),
          Collections.emptyList());
    }
  }

  static void handleWebBasedResult(
          SDKWebResponse sdkWebResponse,
          BillingFlowParams billingFlowParams,
          PurchasesUpdatedListener purchaseFinishedListener
  ) {
    if (sdkWebResponse.getResponseCode() == null) {
      logError("No response code returned on Web Result.");
      purchaseFinishedListener.onPurchasesUpdated(
              ResponseCode.ERROR.getValue(),
              Collections.emptyList()
      );
      return;
    }

    if (sdkWebResponse.getResponseCode() == ResponseCode.OK.getValue()) {
      WalletUtils.getSdkAnalytics().sendPurchaseStatusEvent("success", getResponseDesc(sdkWebResponse.getResponseCode()));
      logDebug("Successful resultcode from purchase activity.");
      logDebug("OrderId: " + sdkWebResponse.getOrderId());
      logDebug("PurchaseToken: " + sdkWebResponse.getPurchaseToken());

      if (sdkWebResponse.getPurchaseToken() == null || sdkWebResponse.getOrderId() == null) {
        logError("BUG: either OrderId or PurchaseToken is null.");
        purchaseFinishedListener.onPurchasesUpdated(
                ResponseCode.ERROR.getValue(),
                Collections.emptyList()
        );

        return;
      }

        try {
          Purchase purchase = sdkWebResponse.toPurchase(billingFlowParams);

          List<Purchase> purchases = new ArrayList<>();
          purchases.add(purchase);

          SendSuccessfulPurchaseResponseEvent.Companion.invoke(purchase);
          purchaseFinishedListener.onPurchasesUpdated(sdkWebResponse.getResponseCode(), purchases);
        } catch (Exception e) {
          e.printStackTrace();
          purchaseFinishedListener.onPurchasesUpdated(
                  ResponseCode.ERROR.getValue(),
                  Collections.emptyList()
          );
          logError("Failed to parse purchase data.");
          return;
        }
        return;
      }

    logAndSendAnalyticsForUnsuccessfulResult(sdkWebResponse.getResponseCode());
    purchaseFinishedListener.onPurchasesUpdated(
            sdkWebResponse.getResponseCode(),
            Collections.emptyList()
    );
  }

  private static int getResponseCodeFromIntent(Intent i) {
    Object o = i.getExtras().get(RESPONSE_CODE);
    if (o == null) {
      logError("Intent with no response code, assuming OK (known issue)");
      return ResponseCode.OK.getValue();
    } else if (o instanceof Integer) {
      return (Integer) o;
    } else if (o instanceof Long) {
      return (int) ((Long) o).longValue();
    } else {
      logError("Unexpected type for intent response code.");
      logError(o.getClass()
          .getName());
      throw new RuntimeException("Unexpected type for intent response code: " + o.getClass()
          .getName());
    }
  }

  private static String getObjectFromJson(JSONObject data, String objectId) {
    return data.optString(objectId);
  }

  private static void logAndSendAnalyticsForUnsuccessfulResult(Integer responseCode) {
    SdkAnalytics sdkAnalytics = WalletUtils.getSdkAnalytics();

    if (responseCode == ResponseCode.USER_CANCELED.getValue()) {
      logDebug("Purchase canceled - Response: " + getResponseDesc(responseCode));
      sdkAnalytics.sendPurchaseStatusEvent("user_canceled", getResponseDesc(responseCode));
    } else {
      logError("Purchase failed. Response code: " + getResponseDesc(
              responseCode));
      sdkAnalytics.sendPurchaseStatusEvent("error", getResponseDesc(responseCode));
    }
  }

  private static String getResponseDesc(int code) {
    String[] iab_msgs = ("0:OK/1:User Canceled/2:Unknown/"
        + "3:Billing Unavailable/4:Item unavailable/"
        + "5:Developer Error/6:Error/7:Item Already Owned/"
        + "8:Item not owned").split("/");
    String[] iabhelper_msgs = ("0:OK/-1001:Remote exception during initialization/"
        + "-1002:Bad response received/"
        + "-1003:Purchase signature verification failed/"
        + "-1004:Send intent failed/"
        + "-1005:User cancelled/"
        + "-1006:Unknown purchase response/"
        + "-1007:Missing token/"
        + "-1008:Unknown error/"
        + "-1009:Subscriptions not available/"
        + "-1010:Invalid consumption attempt").split("/");

    if (code <= -1000) {
      int index = -1000 - code;
      if (index < iabhelper_msgs.length) {
        return iabhelper_msgs[index];
      } else {
        return code + ":Unknown IAB Helper Error";
      }
    } else if (code < 0 || code >= iab_msgs.length) {
      return code + ":Unknown";
    } else {
      return iab_msgs[code];
    }
  }
}

File: android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/ApplicationUtils.kt
package com.appcoins.sdk.billing

import android.app.Activity
import android.content.Intent
import android.util.Base64
import com.appcoins.sdk.billing.CatapultAppcoinsBilling.BillingResponseCode
import com.appcoins.sdk.billing.helpers.BillingResultHelper
import com.appcoins.sdk.billing.usecases.mmp.SendSuccessfulPurchaseResponseEvent
import com.appcoins.sdk.billing.utils.AppcoinsBillingConstants.INAPP_DATA_SIGNATURE
import com.appcoins.sdk.billing.utils.AppcoinsBillingConstants.INAPP_PURCHASE_DATA
import com.appcoins.sdk.billing.utils.AppcoinsBillingConstants.RESPONSE_CODE
import com.appcoins.sdk.core.analytics.SdkAnalytics
import com.appcoins.sdk.core.analytics.SdkAnalyticsUtils
import com.appcoins.sdk.core.logger.Logger.logDebug
import com.appcoins.sdk.core.logger.Logger.logError
import com.appcoins.sdk.core.logger.Logger.logInfo
import com.appcoins.sdk.core.security.PurchasesSecurityHelper
import org.json.JSONObject

internal object ApplicationUtils {

    @Suppress("CyclomaticComplexMethod", "LongMethod")
    @JvmStatic
    fun handleActivityResult(resultCode: Int, data: Intent?, purchaseFinishedListener: PurchasesUpdatedListener) {
        val sdkAnalytics = SdkAnalyticsUtils.sdkAnalytics

        if (data == null) {
            handleDataNull(purchaseFinishedListener, sdkAnalytics)
            return
        }

        val responseCode = getResponseCodeFromIntent(data)
        val purchaseData = data.getStringExtra(INAPP_PURCHASE_DATA)
        val dataSignature = data.getStringExtra(INAPP_DATA_SIGNATURE)

        if (resultCode == Activity.RESULT_OK && responseCode == BillingResponseCode.OK) {
            handleSuccessfulResult(
                responseCode,
                data,
                purchaseData,
                dataSignature,
                purchaseFinishedListener,
                sdkAnalytics
            )
        } else if (resultCode == Activity.RESULT_OK) {
            // result code was OK, but in-app billing response was not OK.
            handleFailureBillingResult(responseCode, data, purchaseFinishedListener, sdkAnalytics)
        } else if (resultCode == Activity.RESULT_CANCELED) {
            handleCanceledResult(responseCode, data, purchaseFinishedListener, sdkAnalytics)
        } else {
            handleUnknownFailureResult(resultCode, responseCode, data, purchaseFinishedListener, sdkAnalytics)
        }
    }

    private fun handleSuccessfulResult(
        responseCode: Int,
        data: Intent,
        purchaseData: String?,
        dataSignature: String?,
        purchaseFinishedListener: PurchasesUpdatedListener,
        sdkAnalytics: SdkAnalytics
    ) {
        logInfo("Successful ResultCode from Purchase.")
        logDebug("Purchase data: $purchaseData")
        logDebug("Data signature: $dataSignature")
        logDebug("Extras: " + data.extras)

        if (purchaseData == null || dataSignature == null) {
            handlePurchaseDataNull(purchaseFinishedListener, sdkAnalytics, data)
            return
        }

        if (PurchasesSecurityHelper.verifyPurchase(purchaseData, Base64.decode(dataSignature, Base64.DEFAULT))) {
            try {
                val purchaseDataJSON = JSONObject(purchaseData)
                val accountIdentifiers =
                    getObjectFromJson(purchaseDataJSON, "obfuscatedExternalAccountId")?.let { AccountIdentifiers(it) }
                val purchase =
                    Purchase(
                        accountIdentifiers,
                        getObjectFromJson(purchaseDataJSON, "developerPayload"),
                        getObjectFromJson(purchaseDataJSON, "orderId", ""),
                        purchaseData,
                        getObjectFromJson(purchaseDataJSON, "packageName", ""),
                        listOf(getObjectFromJson(purchaseDataJSON, "productId", "")),
                        Integer.decode(getObjectFromJson(purchaseDataJSON, "purchaseState", "0")),
                        getObjectFromJson(purchaseDataJSON, "purchaseTime", "0").toLong(),
                        getObjectFromJson(purchaseDataJSON, "purchaseToken", ""),
                        dataSignature,
                        getObjectFromJson(purchaseDataJSON, "isAutoRenewing", "false").toBoolean()
                    )

                val purchases: MutableList<Purchase> = ArrayList()
                purchases.add(purchase)
                SendSuccessfulPurchaseResponseEvent.invoke(purchase)
                sdkAnalytics.sendPurchaseResultEvent(
                    responseCode,
                    purchase.purchaseToken,
                    purchase.products.first()
                )
                purchaseFinishedListener.onPurchasesUpdated(
                    BillingResult.newBuilder().setResponseCode(responseCode).build(),
                    purchases
                )
                logInfo("Purchase result successfully sent.")
            } catch (e: Exception) {
                logError("Failed to parse purchase data: $e")
                sdkAnalytics.sendPurchaseResultEvent(
                    responseCode = BillingResponseCode.ERROR,
                    failureMessage = "Purchase failed with parsing error."
                )
                purchaseFinishedListener
                    .onPurchasesUpdated(
                        BillingResult.newBuilder().setResponseCode(BillingResponseCode.ERROR).build(),
                        emptyList()
                    )
            }
        } else {
            handleSignatureVerificationFailed(purchaseFinishedListener, sdkAnalytics)
        }
    }

    private fun handleDataNull(purchaseFinishedListener: PurchasesUpdatedListener, sdkAnalytics: SdkAnalytics) {
        logError("Null data in IAB activity result.")
        sdkAnalytics.sendPurchaseResultEvent(
            responseCode = BillingResponseCode.ERROR,
            failureMessage = "Null data in IAB activity result."
        )
        purchaseFinishedListener.onPurchasesUpdated(
            BillingResult.newBuilder().setResponseCode(BillingResponseCode.ERROR).build(),
            emptyList()
        )
    }

    private fun handlePurchaseDataNull(
        purchaseFinishedListener: PurchasesUpdatedListener,
        sdkAnalytics: SdkAnalytics,
        data: Intent
    ) {
        logError("BUG: either purchaseData or dataSignature is null.")
        logDebug("Extras: " + data.extras)
        sdkAnalytics.sendPurchaseResultEvent(
            responseCode = BillingResponseCode.ERROR,
            failureMessage = "Either purchaseData or dataSignature is null."
        )
        purchaseFinishedListener.onPurchasesUpdated(
            BillingResult.newBuilder().setResponseCode(BillingResponseCode.ERROR).build(),
            emptyList()
        )
    }

    private fun handleSignatureVerificationFailed(
        purchaseFinishedListener: PurchasesUpdatedListener,
        sdkAnalytics: SdkAnalytics
    ) {
        logError("Signature verification failed.")
        sdkAnalytics.sendPurchaseResultEvent(
            responseCode = BillingResponseCode.DEVELOPER_ERROR,
            failureMessage = "Signature verification failed."
        )
        purchaseFinishedListener.onPurchasesUpdated(
            BillingResultHelper.buildBillingResult(
                BillingResponseCode.DEVELOPER_ERROR,
                BillingResultHelper.ERROR_TYPE_INVALID_PUBLIC_KEY
            ),
            emptyList()
        )
    }

    private fun handleFailureBillingResult(
        responseCode: Int,
        data: Intent,
        purchaseFinishedListener: PurchasesUpdatedListener,
        sdkAnalytics: SdkAnalytics
    ) {
        logError(
            "Result code was OK but in-app billing response was not OK: " +
                getResponseDesc(responseCode)
        )
        logDebug("Bundle: $data")
        sdkAnalytics.sendPurchaseResultEvent(
            responseCode = responseCode,
            failureMessage = "Result code was OK but in-app billing response was not OK."
        )
        purchaseFinishedListener.onPurchasesUpdated(
            BillingResult.newBuilder().setResponseCode(responseCode).build(),
            emptyList()
        )
    }

    private fun handleCanceledResult(
        responseCode: Int,
        data: Intent,
        purchaseFinishedListener: PurchasesUpdatedListener,
        sdkAnalytics: SdkAnalytics
    ) {
        logInfo("Purchase canceled - Response: " + getResponseDesc(responseCode))
        logDebug("Bundle: $data")
        sdkAnalytics.sendPurchaseResultEvent(responseCode = BillingResponseCode.USER_CANCELED)
        purchaseFinishedListener.onPurchasesUpdated(
            BillingResult.newBuilder().setResponseCode(BillingResponseCode.USER_CANCELED).build(),
            emptyList()
        )
    }

    private fun handleUnknownFailureResult(
        resultCode: Int,
        responseCode: Int,
        data: Intent,
        purchaseFinishedListener: PurchasesUpdatedListener,
        sdkAnalytics: SdkAnalytics
    ) {
        logError(
            "Purchase failed. Result code: $resultCode. Response: " +
                getResponseDesc(responseCode)
        )
        logDebug("Bundle: $data")
        sdkAnalytics.sendPurchaseResultEvent(
            responseCode = responseCode,
            failureMessage = "Purchase failed. Result code: $resultCode."
        )
        purchaseFinishedListener.onPurchasesUpdated(
            BillingResult.newBuilder().setResponseCode(BillingResponseCode.ERROR).build(),
            emptyList()
        )
    }

    private fun getResponseCodeFromIntent(intent: Intent): Int =
        intent.getIntExtra(RESPONSE_CODE, BillingResponseCode.ERROR)

    private fun getObjectFromJson(data: JSONObject, objectId: String): String? =
        data.optString(objectId).takeIf { it.isNotEmpty() }

    private fun getObjectFromJson(data: JSONObject, objectId: String, defaultValue: String): String =
        data.optString(objectId).takeIf { it.isNotEmpty() } ?: defaultValue

    private fun getResponseDesc(code: Int): String {
        val iabMsgs = (
            "0:OK/1:User Canceled/2:Unknown/" +
                "3:Billing Unavailable/4:Item unavailable/" +
                "5:Developer Error/6:Error/7:Item Already Owned/" +
                "8:Item not owned"
            ).split("/".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        val iabHelperMsgs = (
            "0:OK/-1001:Remote exception during initialization/" +
                "-1002:Bad response received/" +
                "-1003:Purchase signature verification failed/" +
                "-1004:Send intent failed/" +
                "-1005:User cancelled/" +
                "-1006:Unknown purchase response/" +
                "-1007:Missing token/" +
                "-1008:Unknown error/" +
                "-1009:Subscriptions not available/" +
                "-1010:Invalid consumption attempt"
            ).split("/".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()

        return if (code <= minimumErrorLevel) {
            val index = minimumErrorLevel - code
            if (index < iabHelperMsgs.size) {
                iabHelperMsgs[index]
            } else {
                "$code:Unknown IAB Helper Error"
            }
        } else if (code < startingErrorCode || code >= iabMsgs.size) {
            "$code:Unknown"
        } else {
            iabMsgs[code]
        }
    }

    private const val startingErrorCode = 0
    private const val minimumErrorLevel = -1000
}
